utilities/blob_db/blob_compaction_filter.cc

```diff
@@ -11,89 +11,273 @@
+    const Slice& key, const Slice& existing_value,
+    std::string* new_value) const {
+  assert(new_value);
+
+  const BlobDBImpl* const blob_db_impl = context_gc_.blob_db_impl;
+  (void)blob_db_impl;
+
+  assert(blob_db_impl);
+  assert(blob_db_impl->bdb_options_.enable_garbage_collection);
 
-  ~BlobIndexCompactionFilter() override {
-    RecordTick(statistics_, BLOB_DB_BLOB_INDEX_EXPIRED_COUNT, expired_count_);
-    RecordTick(statistics_, BLOB_DB_BLOB_INDEX_EXPIRED_SIZE, expired_size_);
-    RecordTick(statistics_, BLOB_DB_BLOB_INDEX_EVICTED_COUNT, evicted_count_);
-    RecordTick(statistics_, BLOB_DB_BLOB_INDEX_EVICTED_SIZE, evicted_size_);
+  BlobIndex blob_index;
+  const Status s = blob_index.DecodeFrom(existing_value);
+  if (!s.ok()) {
+    return false;
   }
 
-  const char* Name() const override { return "BlobIndexCompactionFilter"; }
+  if (blob_index.IsInlined()) {
+    return false;
+  }
 
-  // Filter expired blob indexes regardless of snapshots.
-  bool IgnoreSnapshots() const override { return true; }
+  if (blob_index.HasTTL()) {
+    return false;
+  }
 
-  Decision FilterV2(int /*level*/, const Slice& key, ValueType value_type,
-                    const Slice& value, std::string* /*new_value*/,
-                    std::string* /*skip_until*/) const override {
-    if (value_type != kBlobIndex) {
-      return Decision::kKeep;
-    }
-    BlobIndex blob_index;
-    Status s = blob_index.DecodeFrom(value);
-    if (!s.ok()) {
-      // Unable to decode blob index. Keeping the value.
-      return Decision::kKeep;
-    }
-    if (blob_index.HasTTL() && blob_index.expiration() <= current_time_) {
-      // Expired
-      expired_count_++;
-      expired_size_ += key.size() + value.size();
-      return Decision::kRemove;
-    }
-    if (!blob_index.IsInlined() &&
-        blob_index.file_number() < context_.next_file_number &&
-        context_.current_blob_files.count(blob_index.file_number()) == 0) {
-      // Corresponding blob file gone. Could have been garbage collected or
-      // evicted by FIFO eviction.
-      evicted_count_++;
-      evicted_size_ += key.size() + value.size();
-      return Decision::kRemove;
-    }
-    if (context_.fifo_eviction_seq > 0 && blob_index.HasTTL() &&
-        blob_index.expiration() < context_.evict_expiration_up_to) {
-      // Hack: Internal key is passed to BlobIndexCompactionFilter for it to
-      // get sequence number.
-      ParsedInternalKey ikey;
-      bool ok = ParseInternalKey(key, &ikey);
-      // Remove keys that could have been remove by last FIFO eviction.
-      // If get error while parsing key, ignore and continue.
-      if (ok && ikey.sequence < context_.fifo_eviction_seq) {
-        evicted_count_++;
-        evicted_size_ += key.size() + value.size();
-        return Decision::kRemove;
-      }
-    }
-    return Decision::kKeep;
+  if (blob_index.file_number() >= context_gc_.cutoff_file_number) {
+    return false;
+  }
+
+  if (!OpenNewBlobFileIfNeeded()) {
```

Comment:
We are creating a new file per compaction, which might generate many small files. I think it's fine as although some files might be small. total number of files are bounded. Maybe add a comment here to explain that?